napari/utils/misc.py

```diff
@@ -181,7 +181,12 @@ def ensure_sequence_of_iterables(obj, length: Optional[int] = None):
     Out[4]: repeat(None)
     """
 
-    if obj is not None and is_sequence(obj) and is_iterable(obj[0]):
+    if (
+        obj is not None
+        and is_sequence(obj)
+        and len(obj) > 0
```

Comment:
I added this line because serialization was failing for `clipping_planes`. Apparently, until now the case of "empty list as argument" wasn't considered. Hopefully I didn't break anything with this (tests pass, but I'd like confirmation from others that this makes sense).

Reply:
@tlambert03 any further thoughts here?

Reply:
actually, this is https://github.com/napari/napari/issues/3079  (So you could mark this as closing that issue when you fix it).  copied from my comment there:

> I think the behavior for an empty list should be to return an empty list (unless length is provided and not 0, in which case it should be a valueError

so rather than the current implementation (which will return `repeat([])` when provided an empty list), let's just return [] .... _unless_ length is not None, in which case it should assert the length matches

Reply:
I think we have a problem here: as far as I can tell, `repeat([])` *is* the right value in the clipping plane case (`[]` is the argument value, and it should be repeated for each channel). However, as you said in #3079, sometimes `[]` may mean an *empty* sequence of iterables. How do we solve this ambiguity?

Reply:
I feel like that should be up to the consumer of this function.  `[]` is a valid sequence of iterables (of len 0), can you show me where the clipping plane is using this function?

Reply:
If you pass `clipping_planes=[]` (a valid agument) to `Image` in multi-channel mode (let's say with 3 channels), `ensure_sequence_of_iterables` will be called. If the function returns simply `[]` instead of `repeat([])`, all sorts of `IndexError`s pop up, trying to acces the 3 expected elements of the return value.

What are the cases where `[]` interpreted as an empty sequence of iterables is desirable? (0 channels?)

Reply:
just curious, why does `clipping_planes=[]` have to be a valid argument in multi-channel mode?  Kinda seems like `None` is how we're doing "null values" with all the other things.

Reply:
i don't have a terribly strong opinion here... so if you want to just make it `repeat([])` and move on with your life, that's fine.

when using multi-channel mode, `ensure_sequence_of_iterables` will be called with a `length` parameter, and `[]` would _not_ pass that requirement there, so i do think it feels a little bit odd for a valid null value with a specified length.  but since that function is mostly only used for `add_image` anyway, i guess we can make it be what that consumer needs?

Reply:
I think `[]` should be a valid argument because it is valid in single-channel mode, and having inconsistent behaviour would be different from all the other layer parameters (I think?). `None` *is* valid as well. I guess I could change it to use/accept only `None` when the `PlaneList` is empty? That seems a bit counterintuitive to me, but it would solve this issue.

> when using multi-channel mode, ensure_sequence_of_iterables will be called with a length parameter, and [] would not pass that requirement there

Yeah, I noticed that as well. My current solution (WIP) is to add `repeat_empty` as a parameter to the function, default to false. This way, the behaviour stays as in main and the function can be usel elsewhere as previously indended. However, in the multichannel call, we can set `repeat_empty` to true and get this different behaviour. This plays well with `length`:
```python
In [37]: ensure_sequence_of_iterables([])
Out[37]: []

In [38]: ensure_sequence_of_iterables([], length=1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-38-e8e5de8d471d> in <module>
----> 1 ensure_sequence_of_iterables([], length=1)

~/git/napari/napari/utils/misc.py in ensure_sequence_of_iterables(obj, length, repeat_empty)
    194             if (len(obj) == 0 and not repeat_empty) or len(obj) > 0:
    195                 # sequence of iterables of wrong length
--> 196                 raise ValueError(
    197                     trans._(
    198                         "length of {obj} must equal {length}",

ValueError: length of [] must equal 1

In [39]: ensure_sequence_of_iterables([], repeat_empty=True)
Out[39]: repeat([])

In [40]: ensure_sequence_of_iterables([], length=1, repeat_empty=True)
Out[40]: repeat([])
```

What do you think?

Reply:
I also have doubts about this:

https://github.com/napari/napari/blob/aa815ea0430232cb91f7d8111da68c5a880841d3/napari/utils/_tests/test_misc.py#L31-L32

The test seems to imply that this is intended behaviour. Why is is so? This way `((0, 1), None, None)` is treated differently from `(None, (0, 1), None)`, which seems wrong to me. As far as I can tell, both of these tests should result in `repeat((...))`, as they are not sequences of iterables.

I currently changed the function to use `if all(is_iterable(el) for el in obj)` rather than `if len(obj) > 0 and is_iterable(obj[0])`, which results in what I expected, but maybe I'm missing something.